src/item_group.cpp

```diff
@@ -314,7 +314,12 @@ void Item_modifier::modify( item &new_item ) const
             new_item.charges = std::max( 1, ch );
         } else if( new_item.is_tool() ) {
             if( !new_item.magazine_default().is_null() ) {
-                item mag( new_item.magazine_default() );
+                auto magset = new_item.magazine_compatible();
+                auto iter = magset.begin();
+                //make the chance of spawning default magazine slightly lower than others
+                int rng = ( magset.size() * 2 - 1 - ( rand() % ( magset.size() * 2 - 1 ) ) ) / 2;
```

Comment:
```suggestion
                const int rng = rng( 0, magset.size() - 1 );
```
Your original code is most cryptic.

Or even just this:
```suggestion
                item( random_entry( magset ) );
```

Reply:
The purpose of that equation is that I want to make the spawn rate of default magazine (the first entry of that set, atomic for battries) lower than others, It looks bad, but is easy to adjust, I can change the number `2` to `n` in order to make the possibility of spawn default magazine `1/n` lower than others. 
Maybe I should do something similiar but exlusively for batteries? Anyway atomic batteries should be rarer.

Reply:
That might be a bit brittle. Who says the atomic batteries are always the default and at the front? And if atomic batteries are so rare, why are they the default anyway?

Reply:
I don't why, but with my tests the set of compatible magazines is alwyas in alphabetical order. Could you confirm or refute it,? Thank you.

Reply:
>  the set of compatible magazines is always in alphabetical order

Probably because it's stored in a std::set or something alike. So yes, it will always have that order. But that makes it just unbalanced. You're reducing the chance for a specific type of magazine just because it's sorted first (by its type id).

What if a modder adds a magazine with and id that is sorted even before the atomic battery. The atomic magazine would than be used with the default spawn.

And for other items that don't have the atomic battery as possible magazine, it will just reduce the spawn rate for an arbitrary magazine type.

Maybe this should be done via a dedicated item group (which can have dedicated probabilities for each entry).

--

Anyway, if you're only concerned about the atomic battery, what about an explicit test: if chosen magazine is atomic and x-in-y-chance, then randomly chose an magazine again (hopefully another one).

Reply:
I think need to add some `default_magazine` var, and when you read json, first value in compatible magazines list should be cached to this variable, and use `default_magazine` to spawn magazine in most cases.